src/iop/graduatednd.c

```diff
@@ -778,181 +835,145 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
   const int height = roi_out->height;
-  if(data->density > 0)
+  const float length_base = sinv * (-1.0f + ix * hw_inv) + cosv - 1.0f + offset;
+  const float length_inc = sinv * hw_inv * filter_hardness;
+  const float density = data->density;
+  // preload data into aligned storage; the compiler will optimize
+  // these into registers when it vectorizes
+  const dt_aligned_pixel_t color = { data->color[0], data->color[1], data->color[2], data->color[3] };
+  const dt_aligned_pixel_t color1 = { data->color1[0], data->color1[1], data->color1[2], data->color1[3] };
+  const dt_aligned_pixel_t zero = { 0.0f, 0.0f, 0.0f, 0.0f };
+
+  if(density > 0)
   {
 #ifdef _OPENMP
 #pragma omp parallel for default(none) \
-    dt_omp_firstprivate(ch, cosv, data, filter_hardness, hh_inv, hw_inv, \
-                        ivoid, ix, iy, offset, ovoid, height, width, sinv) \
+    dt_omp_firstprivate(density, color, color1, zero, filter_hardness, cosv_hh_inv, \
+                        ivoid, ix, iy, offset, ovoid, height, width, sinv, length_inc, \
+                        length_base)  \
     schedule(static)
 #endif
     for(int y = 0; y < height; y++)
     {
-      const size_t k = (size_t)width * y * ch;
+      const size_t k = (size_t)4 * width * y;
       const float *const restrict in = (float *)ivoid + k;
       float *const restrict out = (float *)ovoid + k;
 
-      float length = (sinv * (-1.0 + ix * hw_inv) - cosv * (-1.0 + (iy + y) * hh_inv) - 1.0 + offset)
-                     * filter_hardness;
-      const float length_inc = sinv * hw_inv * filter_hardness;
+      float length = (length_base - (iy + y) * cosv_hh_inv) * filter_hardness;
+      const dt_aligned_pixel_t counts = { 0.0f, 1.0f, 2.0f, 3.0f };
 
-      for(int x = 0; x < width; x++)
+      // process pixels four at a time so that the compiler can
+      // vectorize the density computation
+      for(int x = 0; x+3 < width; x += 4)
       {
-        const float density = compute_density(data->density, length);
-
-        #ifdef _OPENMP
-        #pragma omp simd aligned(in, out : 16)
-        #endif
-        for(int l = 0; l < 4; l++)
+        dt_aligned_pixel_t curr_density;
+        dt_aligned_pixel_t lengths;
+        for_four_channels(i)
         {
-          out[ch*x+l] = MAX(0.0f, (in[ch*x+l] / (data->color[l] + data->color1[l] * density)));
+          lengths[i] = length + counts[i] * length_inc;
+          curr_density[i] = _compute_density(density, lengths[i]);
         }
-        length += length_inc;
+        for(int i = 0; i < 4; i++)
+        {
+          dt_aligned_pixel_t res;	// the compiler will optimize this into a register
+          for_each_channel(l, aligned(in : 16))
+          {
+            res[l] = MAX(zero[l], (in[4*(x+i)+l] / (color[l] + color1[l] * curr_density[i])));
+          }
+          // use streaming writes to eliminate the memory reads from loading cache lines
+          copy_pixel_nontemporal(out + 4*(x+i), res);
```

Comment:
Here we are outside of an SSE section.

Reply:
I've moved the use of _mm_stream_ps into the definition of copy_pixel_nontemporal in darktable.h so there are no more direct  references to SSE in graduatednd.c.